File: 8.Ros/Ros2/net/src/publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "example_interfaces/msg/string.hpp"
#include <chrono>
#include <iostream>

using namespace std::chrono_literals;

class MyNode : public rclcpp::Node
{
private:
    rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
public:
    MyNode() : Node("publisher")
    {
        publisher_ = this->create_publisher<example_interfaces::msg::String>("/my_topic", 10);
        timer_ = this->create_wall_timer(500ms, std::bind(&MyNode::publish_message, this));
    }
    void publish_message()
    {
        auto message = example_interfaces::msg::String();
        message.data = "Hello, this is a custom message!";
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
        publisher_->publish(message);
    }
    ~MyNode()
    {
        RCLCPP_INFO(this->get_logger(),"destructor");
    }
    
};

int main()
{

    rclcpp::init(0, nullptr);
    auto node = std::make_shared<MyNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: 8.Ros/Ros2/soso/soso/publisher.py
import rclpy
from std_msgs.msg import String




def main():
    rclpy.init()
    node = rclpy.create_node('soso_publisher')
    pub = node.create_publisher(String, '/topic', 10)
    msg = String()
    i = 0

    def timer_callback():
        nonlocal i
        msg.data = "hello ismail %d" % i
        node.get_logger().info('Publishing: "%s"' % msg.data)
        i += 1
        pub.publish(msg)

    timer = node.create_timer(0.5, timer_callback)
    rclpy.spin(node)
    node.destroy_timer(timer)    
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()